app/org/sagebionetworks/bridge/hibernate/HibernateAccountDao.java

```diff
@@ -248,10 +270,69 @@ public void deleteReauthToken(AccountId accountId) {
             hibernateAccount.setReauthTokenHash(null);
             hibernateAccount.setReauthTokenAlgorithm(null);
             hibernateAccount.setReauthTokenModifiedOn(null);
-            hibernateHelper.update(hibernateAccount);
+            try { 
+                hibernateHelper.update(hibernateAccount);
+            } catch(PersistenceException pe) {
+                throw convertPersistenceException(pe, hibernateAccount.getId());
+            }            
         }
     }
     
+    private RuntimeException convertPersistenceException(PersistenceException exception, String userId) {
+        // The sequence of type-checking and unwrapping of this exception is significant as unfortunately, 
+        // the hierarchy of wrapped exceptions is very specific. 
+        if (exception instanceof OptimisticLockException) {
+            return new ConcurrentModificationException(
+                    "Account has the wrong version number; it may have been saved in the background.");
+        }
+        if (exception.getCause() instanceof org.hibernate.exception.ConstraintViolationException) {
+            // The specific error message is buried in the root MySQLIntegrityConstraintViolationException
+            Throwable cause = Throwables.getRootCause(exception);
+            String message = cause.getMessage();
+            if (message != null && userId != null) {
+                Map<String,Object> entityKeys = ImmutableMap.of("userId", userId);
+                
+                if (message.matches("Duplicate entry.*for key 'Accounts-StudyId-ExternalId-Index'")) {
+                    return new EntityAlreadyExistsException(Account.class, "External ID has already been used by another account.", entityKeys);
+                } else if (message.matches("Duplicate entry.*for key 'Accounts-StudyId-Email-Index'")) {
+                    return new EntityAlreadyExistsException(Account.class, "Email address has already been used by another account.", entityKeys);
+                } else if (message.matches("Duplicate entry.*for key 'Accounts-StudyId-Phone-Index'")) {
+                    return new EntityAlreadyExistsException(Account.class, "Phone number has already been used by another account.", entityKeys);
+                }
+            }
+            ConstraintViolationException.Builder cveBuilder = new ConstraintViolationException.Builder();
+            if (message != null) {
+                cveBuilder.withMessage(cause.getMessage());
+            }
+            return cveBuilder.build();
+        }
+        return new BridgeServiceException(exception);
```

Comment:
here, I'm now thinking we should just rethrow the error if not translated. Consistence with non-update calls to HibernateHelper.